cupy/cuda/memory.pyx

```diff
@@ -1021,53 +1028,75 @@ cpdef size_t _bin_index_from_size(size_t size):
     return (size - 1) // ALLOCATION_UNIT_SIZE
 
 
-cdef _gc_isenabled = gc.isenabled
-cdef _gc_disable = gc.disable
-cdef _gc_enable = gc.enable
-
-
-cdef bint _lock_no_gc(lock):
+cdef bint _lock_no_gc(recursive_mutex& lock):
     """Lock to ensure single thread execution and no garbage collection.
 
     Returns:
         bool: Whether GC is disabled.
     """
-    rlock.lock_fastrlock(lock, -1, True)
+    if not lock.try_lock():
+        with nogil:
+            lock.lock()
```

Comment:
I suspect this is unsafe (until we make the memory pool free-threaded-safe), because this will give other threads a chance to call memory pool APIs.